### PR TITLE
Add XamlApplication supporting XAML features for WinUI 3

### LINKED ISSUE #25: Proposal: XamlApplication: Application implementing IXamlMetadataPorvider interface
# Proposal: XamlApplication: Application implementing IXamlMetadataPorvider interface

## Summary

XamlApplication is a Application implementing IXamlMetadataProvider interface and loading WinUI theme. This API provides an easy way to create an application integrating XAML.

```idl
runtimeclass XamlApplication : Microsoft.UI.Xaml.Application, Microsoft.UI.Xaml.Markup.IXamlMetadataProvider
```

## Rationale

* WinUI 3 apps need Application implementing IXamlMetadataProvider.
* This API makes a code simpler.

## Sample code

```cpp
// Application and IXamlMetadataProvider
struct App : public ApplicationT<App, IXamlMetadataProvider>
{
public:
	void OnLaunched(LaunchActivatedEventArgs const&)
	{
		Resources().MergedDictionaries().Append(XamlControlsResources());
		...
	}
	IXamlType GetXamlType(TypeName const& type)
	{
		return provider.GetXamlType(type);
	}
	IXamlType GetXamlType(hstring const& fullname)
	{
		return provider.GetXamlType(fullname);
	}
	com_array<XmlnsDefinition> GetXmlnsDefinitions()
	{
		return provider.GetXmlnsDefinitions();
	}
private:
	...
	XamlControlsXamlMetaDataProvider provider;
};

// XamlApplication
struct App : public XamlApplicationT<App>
{
public:
	void OnLaunched(LaunchActivatedEventArgs const&)
	{
		Resources().MergedDictionaries().Append(XamlControlsResources());
		...
	}
};
```

### PR DESCRIPTION
**Pull request by sotanakamura**
This PR adds XamlApplication class supporting XAML features for WinUI 3, which enable us to use WinUI 3 controls with fluent design. This code has an error on reference counting now, but works.  I think you can find the bug more easily than me, please review it.
fix #25

### CODE REVIEW

File: win32more/xaml.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        return self.comptr.QueryInterface(riid, ppvObject)
+
+    @WINFUNCTYPE(UInt32, c_void_p)
+    def AddRef(this):
+        self = cast(this, POINTER(XamlApplicationImpl)).contents
+        return self.comptr.AddRef()
+
+    @WINFUNCTYPE(UInt32, c_void_p)
+    def Release(this):
+        self = cast(this, POINTER(XamlApplicationImpl)).contents
```
Comment (ynkdir) [inline win32more/xaml.py:61]: `this` is `pointer(lpvtbl2)` when invoked as IXamlMetadataProvider.
In this case `cast(this POINTER(XamlApplicationImpl)` is invalid.

Comment (sotanakamura) [reply win32more/xaml.py:61]: Thank you. I fixed it with a little bit dirty way.

**Comment by sotanakamura** [comment]
I fixed the reference counting bug, but it still have a error (the program exit with code 1).

**Comment by ynkdir** [comment]
> It failed to get a xaml type byfull name.

I can't see this error.

In my environment, exit code is 0xC0000005.
It might be fixed by `_inner_interface.Release()`.

```diff
diff --git a/win32more/xaml.py b/win32more/xaml.py
index d6a01069..ac0688d2 100644
--- a/win32more/xaml.py
+++ b/win32more/xaml.py
@@ -163,6 +163,7 @@ class XamlApplication(IApplicationOverrides):
         self._refcount -= 1
         if self._refcount == 0:
             self._comobj.comptr = None
+            self._inner_interface.Release()
             del self._keep_reference_in_python_world_[id(self)]
         return self._refcount
 
@@ -196,9 +197,19 @@ class XamlApplication(IApplicationOverrides):
         return S_OK
 
     @classmethod
-    def Start(cls, app):
+    def Start(cls, appcls):
         hr = MddBootstrapInitialize2(0x00010004, "", PACKAGE_VERSION(Version=0x0FA0041900750000), MddBootstrapInitializeOptions_OnNoMatch_ShowUI)
         if FAILED(hr):
             raise WinError(hr)
-        Application.Start(lambda params: app() and None)
+
+        app = None
+
+        def init(params):
+            nonlocal app
+            app = appcls()
+
+        Application.Start(init)
+
+        app.Release()
+
         MddBootstrapShutdown()
```

**Comment by ynkdir** [comment]
Thank you.

> By the way, do we have as function to convert Iinspectable into another winrt type?

I think it's good.

**Pull request merged**